### PR TITLE
feat: implement /loop command with fixed and dynamic scheduling

### PR DESCRIPTION
**Pull request by khaledmoayad**
## Summary

Implements the `/loop` skill with four scheduling modes and enables the cron infrastructure needed for it to work in the open build.

## The Bun constant-folding problem

While testing, we discovered that Bun's constant folder evaluates `feature('AGENT_TRIGGERS')` at bundle time through the `bun:bundle` shim — and the folded value is **cached from a previous build**. This means flipping the flag to `true` in `build.ts` doesn't help: the `require()` blocks inside `if (feature('AGENT_TRIGGERS'))` still compile to `if (false) {}` dead code, so cron tools, `useScheduledTasks`, and loop skill registration are all silently dropped from the bundle.

This is why we need to remove the guards rather than just flip the flag — and why #633 alone won't be enough to activate `/loop`.

## Changes

**Loop skill** (`loop.ts`, `loop.test.ts`):
- Four modes: fixed-interval with prompt, fixed-interval maintenance, dynamic self-pacing, dynamic maintenance (bare `/loop`)
- Interval parsing accepts leading tokens, trailing "every" clauses, and human-readable units
- Maintenance mode reads `.claude/loop.md` or `~/.claude/loop.md`, falls back to built-in prompt

**Infra** (needed due to the Bun bundling issue above):
- `tools.ts`: cron tools always registered; `isEnabled` gates visibility at runtime
- `REPL.tsx`: `useScheduledTasks` always mounted via unconditional import
- `index.ts`: `registerLoopSkill` via static import, called unconditionally
- `prompt.ts`: `isKairosCronEnabled()` returns `true` for non-ant builds without consulting the feature flag

## Test plan

- [x] `bun test src/skills/bundled/loop.test.ts` — all four modes pass
- [x] `node dist/cli.mjs` — `/loop 1m say hi` schedules, fires, and repeats
- [x] bare `/loop` enters dynamic maintenance mode and self-reschedules

### CODE REVIEW

File: src/tools.ts [hunk 1]
```diff
@@ -26,13 +26,11 @@ const SleepTool =
...
-      require('./tools/ScheduleCronTool/CronCreateTool.js').CronCreateTool,
-      require('./tools/ScheduleCronTool/CronDeleteTool.js').CronDeleteTool,
-      require('./tools/ScheduleCronTool/CronListTool.js').CronListTool,
-    ]
-  : []
+const cronTools = [
+  require('./tools/ScheduleCronTool/CronCreateTool.js').CronCreateTool,
+  require('./tools/ScheduleCronTool/CronDeleteTool.js').CronDeleteTool,
+  require('./tools/ScheduleCronTool/CronListTool.js').CronListTool,
+]
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/tools.ts:33]: The `IN_PROCESS_TEAMMATE_ALLOWED_TOOLS` allowlist in `constants/tools.ts` only matters when the `TEAMMEM` flag is on, which is also `false` in the open build. That entire code path is dead here, so there's no runtime inconsistency. The cron tools need to be unconditionally registered so they're available to the main session for `/loop`; gating them back behind `AGENT_TRIGGERS` would break the feature.

File: src/screens/REPL.tsx [hunk 2]
```diff
@@ -4076,21 +4076,13 @@ export function REPL({
...
-      setMessages
-    });
-  }
+  // and session-only /loop runs.
+  const assistantMode = store.getState().kairosEnabled;
+  useScheduledTasks({
+    isLoading,
+    assistantMode,
+    setMessages
+  });
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/screens/REPL.tsx:4085]: The hook needs to be mounted unconditionally so it's ready when `CronCreate` fires during a `/loop` invocation, that's the whole point of this change. The poll loop is lightweight (checks an in-memory flag) and matches upstream behavior. The suggested replacement (`if ("external" === 'ant')`) would disable the hook entirely in open builds, which breaks `/loop`.

File: src/skills/bundled/loop.ts [hunk 3]
```diff
@@ -6,87 +6,239 @@ import {
...
+
+const MAINTENANCE_PROMPT = `Scheduled maintenance loop iteration.
+
+If .claude/loop.md exists, read it and follow it.
+Otherwise, if ~/.claude/loop.md exists, read it and follow it.
+Otherwise:
+- continue any unfinished work from the conversation
+- tend to the current branch's pull request: review comments, failed CI runs, merge conflicts
+- run cleanup passes such as bug hunts or simplification when nothing else is pending
+
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.ts:32]: This follows the same trust model as `.claude/CLAUDE.md`. repo-level config files are read by the assistant and gated by the existing permission system. The maintenance prompt also explicitly states that irreversible actions only proceed when already authorized in the transcript. Restricting to `~/.claude/loop.md` only would make the feature less useful for per-project loop configurations.

File: src/skills/bundled/loop.ts [hunk 4]
```diff
@@ -6,87 +6,239 @@ import {
...
+}
 
-| Interval pattern      | Cron expression     | Notes                                    |
-|-----------------------|---------------------|------------------------------------------|
-| \`Nm\` where N ≤ 59   | \`*/N * * * *\`     | every N minutes                          |
-| \`Nm\` where N ≥ 60   | \`0 */H * * *\`     | round to hours (H = N/60, must divide 24)|
-| \`Nh\` where N ≤ 23   | \`0 */N * * *\`     | every N hours                            |
-| \`Nd\`                | \`0 0 */N * *\`     | every N days at midnight local           |
-| \`Ns\`                | treat as \`ceil(N/60)m\` | cron minimum granularity is 1 minute  |
+const DEFAULT_INTERVAL = '10m'
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.ts]: Will remove in the next commit

**Comment by khaledmoayad** [comment]
@kevincodex1 @Vasanthdev2004 can I get a review on this?

File: src/skills/bundled/loop.ts [hunk 5]
```diff
@@ -6,87 +6,238 @@ import {
...
+
+For the recurring scheduled task, use this exact maintenance prompt body:
+
+\
+\
+\
+${MAINTENANCE_PROMPT}
+\
+\
+\
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.ts]: Fixed — replaced all backslash line-continuation delimiters with `--- BEGIN/END ---` markers throughout both `buildFixedPrompt` and `buildDynamicPrompt`.

File: src/skills/bundled/loop.ts [hunk 6]
```diff
@@ -6,87 +6,238 @@ import {
...
+`
+
+  return `# /loop — fixed recurring interval
+
+The user invoked /loop with a fixed interval.
+
+Requested interval: \
+${parsed.interval}\
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.ts]: Fixed along with the other delimiter issues in the same commit.

File: src/skills/bundled/loop.ts [hunk 7]
```diff
@@ -6,87 +6,238 @@ import {
...
+   - Pin the cron expression to a specific future local-time minute that matches the chosen delay.
+   - Set the scheduled prompt to this exact text so the next iteration stays in dynamic mode:
+
+\
+\
+\
+${reschedulePrompt}
+\
+\
+\
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.ts]: Fixed — now uses `--- BEGIN SCHEDULED PROMPT ---` / `--- END SCHEDULED PROMPT ---` markers.

File: src/skills/bundled/loop.test.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  const skill = getBundledSkills().find(command => command.name === 'loop')
+  const blocks = await skill!.getPromptForCommand('check the deploy', {} as never)
+  const text = (blocks[0] as { text: string }).text
+
+  expect(text).toContain('# /loop — dynamic rescheduling')
+  expect(text).toContain('check the deploy')
+  expect(text).toContain('choose the next delay dynamically between 1 minute and 1 hour')
+  expect(text).toContain('/loop check the deploy')
+})
```
[the remark this answers is not in the document]
Comment (khaledmoayad) [reply src/skills/bundled/loop.test.ts:38]: Added tests for: trailing `every` clause with compact unit (`every 20m`), trailing `every` with word unit (`every 5 minutes`), the `check every PR` non-interval case, hour unit normalization, and a delimiter presence check.

**Review by Vasanthdev2004 — CHANGES_REQUESTED** [review]
## Review: `/loop` command with fixed and dynamic scheduling

Thanks for this — the four-mode `/loop` design (fixed-prompt, fixed-maintenance, dynamic-prompt, dynamic-maintenance) is well thought out, and the interval parsing is clean. The Bun constant-folding problem is a genuine gotcha and the workaround (removing build-time guards so the code gets bundled) makes sense.

CI is green ✅. Tests cover all four modes plus edge cases like `check every PR` not matching the trailing `every` clause — nice.

That said, the partial removal of `feature('AGENT_TRIGGERS')` guards creates an inconsistency that could break cron in non-interactive mode.

---

### 🔴 Blocker: `feature('AGENT_TRIGGERS')` guards removed from 4 files but left in 2 others

This PR removes the `feature('AGENT_TRIGGERS')` guard from:

- ✅ `src/tools.ts` — cron tools always registered
- ✅ `src/screens/REPL.tsx` — `useScheduledTasks` always mounted
- ✅ `src/skills/bundled/index.ts` — `registerLoopSkill` always called
- ✅ `src/tools/ScheduleCronTool/prompt.ts` — `isKairosCronEnabled()` returns `true` for non-ant builds

But **leaves** the guard in:

- ❌ `src/cli/print.ts` lines 365–371 — `cronSchedulerModule`, `cronJitterConfigModule`, `cronGate` are all `null` when `AGENT_TRIGGERS` is off
- ❌ `src/cli/print.ts` line 2705 — cron scheduler is only created when `feature('AGENT_TRIGGERS')` is true AND `isKairosCronEnabled()` — in open builds, both the module-level `require`s are dead AND the scheduler init block is skipped
- ❌ `src/constants/tools.ts` line 85 — cron tool names are only added to `DISALLOWED_ASYNC_TOOLS` when `AGENT_TRIGGERS` is on

**Result:** In the open build after this PR merges:

| Scenario | REPL (interactive) | `-p` / non-interactive |
|---|---|---|
| Cron tools registered | ✅ (always) | ✅ (always) |
| `isKairosCronEnabled()` | ✅ returns true | ✅ returns true |
| Cron scheduler running | ✅ (useScheduledTasks mounts) | ❌ **dead** — `feature('AGENT_TRIGGERS')` is false at bundle time, so `cronSchedulerModule` is `null` and the scheduler is never created |
| `/loop` in `-p` mode | N/A | ❌ **broken** — tools exist, skill is registered, but no scheduler fires them |

The REPL interactive path works fine because `useScheduledTasks` has its own `isKairosCronEnabled()` check and creates the scheduler independently. But the non-interactive path (`print.ts`) is gated by a build-time constant that still evaluates to `false`.

**Fix:** Also remove the `feature('AGENT_TRIGGERS')` guards from `src/cli/print.ts` (lines 365–371, 2705) and `src/constants/tools.ts` (line 85), matching the pattern used in the other files. Or — if the intent is to only support `/loop` in interactive REPL mode for now — document that explicitly so it's not a surprise.

---

### 🟡 Non-blocking: `feature('AGENT_TRIGGERS')` is still referenced in PR description but now dead in 4/6 files

The PR body explains the Bun constant-folding problem clearly. But after the partial guard removal, the term `AGENT_TRIGGERS` is a bit of a zombie — still referenced in code comments and `print.ts`/`constants/tools.ts` guards, but no longer the actual gate for the open build. Worth updating the PR description or adding a code comment summarizing the new gating strategy:

> **Open builds:** `isKairosCronEnabled()` always returns `true` (unless `CLAUDE_CODE_DISABLE_CRON=1`).
> **Ant builds:** `isKairosCronEnabled()` defers to GrowthBook `tengu_kairos_cron`.
> **Build-time `feature('AGENT_TRIGGERS')` is no longer the cron gate** — it's a bundle-time constant that Bun folds. The runtime gate is `isKairosCronEnabled()`.

---

### 🟢 Loop skill implementation — clean

- `parseLoopArgs()` is well-structured with clear priority: bare interval → leading interval → trailing `every` clause → dynamic.
- `parseTrailingEveryClause` correctly avoids matching `check every PR` because `PR` isn't a time unit — good edge case coverage.
- `normalizeIntervalUnit` supports both short and long forms — nice UX.
- `buildFixedPrompt` and `buildDynamicPrompt` both use `--- BEGIN/END PROMPT ---` delimiters for unambiguous prompt extraction.
- Dynamic mode rescheduling via `recurring: false` + self-referencing `/loop` prompt is an elegant design.
- Maintenance mode with `.claude/loop.md` / `~/.claude/loop.md` fallback + built-in prompt is practical.
- The `isEnabled: isKairosCronEnabled` delegation on the skill is correct — it gates visibility at the skill level, matching the tool pattern.

### 🟢 REPL.tsx change — safe

`useScheduledTasks` has an internal `if (!isKairosCronEnabled()) return` guard at the top of its `useEffect`. Unconditionally mounting it is rules-of-hooks compliant and just defers the gate to runtime. This is exactly the pattern described in the old comment — just without the build-time dead-code wrapper.

### 🟢 Tests — thorough

125 lines covering all 4 modes, edge cases (`check every PR`), human-readable units, and prompt delimiter presence. Good coverage for a prompt-generation skill.

---

### Verdict: **Needs changes** 🔧

One blocker: the partial `AGENT_TRIGGERS` guard removal leaves cron non-functional in `-p` / non-interactive mode. The fix is to also remove the guards from `print.ts` and `constants/tools.ts`, or explicitly document that `/loop` is REPL-only for now.

Once the guard removal is consistent across all 6 files (or the limitation is documented), this is approve-ready. The `/loop` skill implementation itself is solid.

**Comment by khaledmoayad** [comment]
@Vasanthdev2004 great catch — you're right, the guard removal was incomplete.

Fixed in 6cfb665:

- `src/cli/print.ts`: `cronSchedulerModule`, `cronJitterConfigModule`, and `cronGate` are now unconditional requires. The scheduler init block drops the `feature('AGENT_TRIGGERS')` check and delegates entirely to `cronGate.isKairosCronEnabled()` (which returns `true` for non-ant builds unless `CLAUDE_CODE_DISABLE_CRON=1`).
- `src/constants/tools.ts`: Cron tool names are always included in `IN_PROCESS_TEAMMATE_ALLOWED_TOOLS` — no more feature-gated spread.

`/loop` should now work correctly in `-p` / non-interactive mode as well as the REPL.

**Review by Vasanthdev2004 — APPROVED** [review]
## Re-review after requested changes

Reviewed on head `6cfb665`. CI green ✅.

Both blockers from the previous review are now fixed:

### ✅ `src/cli/print.ts` — guards removed

- `cronSchedulerModule`, `cronJitterConfigModule`, `cronGate` are now unconditional `require()`s — no more `feature('AGENT_TRIGGERS')` ternary wrappers
- `runHeadlessStreaming` scheduler init: `if (cronGate.isKairosCronEnabled())` — clean single-gate, no more `feature('AGENT_TRIGGERS') &&` prefix
- Optional chaining removed from `cronJitterConfigModule?.getCronJitterConfig` → `cronJitterConfigModule.getCronJitterConfig` and `cronGate?.isKairosCronEnabled()` → `cronGate.isKairosCronEnabled()` — correct since the modules are always loaded now

### ✅ `src/constants/tools.ts` — cron tools always included

`CRON_CREATE_TOOL_NAME`, `CRON_DELETE_TOOL_NAME`, `CRON_LIST_TOOL_NAME` are now unconditionally spread into `IN_PROCESS_TEAMMATE_ALLOWED_TOOLS`. No more `feature('AGENT_TRIGGERS')` conditional.

### ✅ `isKairosCronEnabled()` — clean two-path gate

The updated logic is simple and correct:
- `CLAUDE_CODE_DISABLE_CRON=1` → always false (local kill switch)
- `USER_TYPE !== 'ant'` → always true (open builds)
- Otherwise → defer to GrowthBook `tengu_kairos_cron` with `true` default

This is the right stratification — open builds get cron unconditionally, Anthropic-internal builds keep the fleet-wide kill switch.

### Result: consistent guard removal across all 6 files

| File | Guard removed? |
|---|---|
| `src/tools.ts` | ✅ (previous commit) |
| `src/screens/REPL.tsx` | ✅ (previous commit) |
| `src/skills/bundled/index.ts` | ✅ (previous commit) |
| `src/tools/ScheduleCronTool/prompt.ts` | ✅ (previous commit) |
| `src/cli/print.ts` | ✅ **fixed in this commit** |
| `src/constants/tools.ts` | ✅ **fixed in this commit** |

Cron now works in both REPL and `-p`/non-interactive mode for open builds. The `/loop` skill implementation remains solid.

---

**Verdict: Approve-ready** ✅

Clean fix, all blockers resolved. The guard removal is now consistent and the runtime gating (`isKairosCronEnabled()`) is the single source of truth.

**Comment by khaledmoayad** [comment]
@kevincodex1 @euxaristia would appreciate a review on this if you have a moment!

**Review by kevincodex1 — APPROVED** [review]
lgtm thanks bro!

**Review by Flo5k5 — COMMENTED** [review]
Nice work on the loop rewrite — the 4-mode parsing with proper code-level extraction is a solid improvement over the "let the LLM figure it out" approach, and the maintenance mode with .claude/loop.md is a great touch.

Re: your latest comment — agreed that removing the guards and using `isKairosCronEnabled()` as the sole runtime gate is the cleaner long-term pattern. Since #657 fixes the build system foundation, I'll land it first so you have a clean base. #633 will become redundant once #621 lands (same guards, different strategy), so I'll close it at that point.

Minor suggestion: in `isKairosCronEnabled()`, the `process.env.USER_TYPE !== 'ant'` check works correctly for the open build but relies on an implicit convention. A one-line comment noting that USER_TYPE is never 'ant' in open builds would help future readers.

Tests are thorough — the "check every PR" edge case is a nice catch. LGTM.
